Common/Arm64Emitter.cpp

```diff
@@ -313,24 +313,40 @@ void ARM64XEmitter::FlushIcache()
 
 void ARM64XEmitter::FlushIcacheSection(u8* start, u8* end)
 {
-	if (cpu_info.sBugs.bExynos8890Invalidation)
-	{
-		// Over invalidate to force this CPU to listen.
-		start = m_startcode + 4096 < start ? start - 4096 : m_startcode;
-		end += 4096;
-	}
-
 #if defined(IOS)
 	// Header file says this is equivalent to: sys_icache_invalidate(start, end - start);
 	sys_cache_control(kCacheFunctionPrepareForExecution, start, end - start);
-#else
-#if (defined(__clang__) && !defined(_M_IX86) && !defined(_M_X64)) || defined(ANDROID)
-	__clear_cache(start, end);
-#else
-#if !defined(_M_IX86) && !defined(_M_X64)
-	__builtin___clear_cache(start, end);
-#endif
-#endif
+#elif !defined(_M_IX86) && !defined(_M_X64)
+	// Code from Dolphin, contributed by the Mono project.
+
+	// Don't rely on GCC's __clear_cache implementation, as it caches
+	// icache/dcache cache line sizes, that can vary between cores on
+	// big.LITTLE architectures.
+	u64 addr, ctr_el0;
+	static size_t icache_line_size = 0xffff, dcache_line_size = 0xffff;
+	size_t isize, dsize;
+
+	__asm__ volatile("mrs %0, ctr_el0" : "=r"(ctr_el0));
+	isize = 4 << ((ctr_el0 >> 0) & 0xf);
+	dsize = 4 << ((ctr_el0 >> 16) & 0xf);
+
+	// use the global minimum cache line size
+	icache_line_size = isize = icache_line_size < isize ? icache_line_size : isize;
+	dcache_line_size = dsize = dcache_line_size < dsize ? dcache_line_size : dsize;
```

Comment:
yes, this is correct (it's _unlikely_ though).

The right way would be to iterate over all available cores on startup and figure out the minimum cache line size. I couldn't figure out a reliable way to do that on Android. Or, since you already have infrastructure for device specific quirks, you could hard code it...